packages/asset-swapper/src/utils/market_operation_utils/index.ts

```diff
@@ -331,31 +372,59 @@ function createBuyPathFromNativeOrders(orders: SignedOrderWithFillableAmounts[])
     return path;
 }
 
-function createPathsFromDexQuotes(dexQuotes: DexSample[][], noConflicts: boolean): Fill[][] {
+function createSellPathsFromDexQuotes(
+    dexQuotes: DexSample[][],
+    ethToOutputAssetRate: BigNumber,
+    opts: GetMarketOrdersOpts,
+): Fill[][] {
+    return createPathsFromDexQuotes(dexQuotes, ethToOutputAssetRate, opts);
+}
+
+function createBuyPathsFromDexQuotes(
+    dexQuotes: DexSample[][],
+    ethToOutputAssetRate: BigNumber,
+    opts: GetMarketOrdersOpts,
+): Fill[][] {
+    return createPathsFromDexQuotes(
+        dexQuotes,
+        // Negated because we try to minimize the output in buys.
+        ethToOutputAssetRate.negated(),
+        opts,
+    );
+}
+
+function createPathsFromDexQuotes(
+    dexQuotes: DexSample[][],
+    ethToOutputAssetRate: BigNumber,
+    opts: GetMarketOrdersOpts,
+): Fill[][] {
     const paths: Fill[][] = [];
     for (const quote of dexQuotes) {
-        // Native orders can be filled in any order, so they're all root nodes.
         const path: Fill[] = [];
         let prevSample: DexSample | undefined;
         // tslint:disable-next-line: prefer-for-of
         for (let i = 0; i < quote.length; i++) {
             const sample = quote[i];
-            if (sample.output.eq(0) || (prevSample && prevSample.output.gte(sample.output))) {
-                // Stop if the output is zero or does not increase.
+            // Stop of the sample has zero output, which can occur if the source
+            // cannot fill the full amount.
+            if (sample.output.isZero()) {
                 break;
             }
             path.push({
-                parent: path.length !== 0 ? path[path.length - 1] : undefined,
-                flags: sourceToFillFlags(sample.source),
-                exclusionMask: noConflicts ? sourceToExclusionMask(sample.source) : 0,
                 input: sample.input.minus(prevSample ? prevSample.input : 0),
                 output: sample.output.minus(prevSample ? prevSample.output : 0),
+                fillPenalty: ZERO_AMOUNT,
+                parent: path.length !== 0 ? path[path.length - 1] : undefined,
+                flags: sourceToFillFlags(sample.source),
+                exclusionMask: opts.noConflicts ? sourceToExclusionMask(sample.source) : 0,
                 fillData: { source: sample.source },
             });
             prevSample = quote[i];
         }
+        // Don't push empty paths.
         if (path.length > 0) {
-            // Don't push empty paths.
+            // Only the first fill in a DEX path incurs a penalty.
+            path[0].fillPenalty = ethToOutputAssetRate.times(opts.fees[path[0].fillData.source] || 0);
```

Comment:
```suggestion
            path[0].fillPenalty = ethToOutputAssetRate.times(opts.fees[path[0].fillData.source] || 0).integerValue();
```

Reply:
I'm kind of leaning towards keeping these full precision since they're just used for comparison.